File: 39-combination-sum/39-combination-sum.cpp
class Solution {
public:
    void helper(vector<int>& candidates, int idx, int target, vector<vector<int>>& output, vector<int> temp) {
        if(target == 0) {
            output.push_back(temp);
            return;
        }
        
        if(target < 0) {
            return;
        }
        
        if(idx >= candidates.size()) {
            return;
        }
        
        temp.push_back(candidates[idx]);
        helper(candidates,idx,target-candidates[idx],output,temp);
        temp.pop_back();
        helper(candidates,idx+1,target,output,temp);
    }
    
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> output;
        
        helper(candidates,0,target,output,{});
        
        return output;
    }
};

File: 39-combination-sum/39-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        output = []
        
        def helper(idx,curSum, li):
            if curSum == 0:
                output.append(li.copy())
                return
            
            if curSum < 0 or idx >= len(candidates):
                return
            
            li.append(candidates[idx])
            helper(idx,curSum-candidates[idx],li)
            li.pop()
            helper(idx+1,curSum,li)
        
        helper(0,target,[])
        
        return output